### ISSUE TITLE
Get keywords with 0 uses

### ISSUE THREAD

**Issue by dmd** [RhetTbull/osxphotos#1226]
In Photos, if you create a keyword, and then delete it, it still stays as a completable keyword, and still shows up in Manage My Keywords.

`osxphotos keywords` however only shows keywords with at least one occurence.

I'd like to get a list of all keywords with zero occurrences, so I can delete them so they don't pollute my completion-space.

**Comment by oPromessa** [comment]
As you probably know but on the Manage My Keywords window when you delete a keyword, it will notify you if there are pics associated with it. 

Not ideal ;) But what I normally do is go through the ones not in the osxphotos keyords list and "confidently" try to delete them.. knowing if there are any pics with that keyword, Photos will notify me.

**Comment by dmd** [comment]
That may work for people who don't use keywords very heavily, but:

```
$ osxphotos keywords |wc -l
Using last opened Photos library: /Users/ddrucker/Pictures/Photos Library.photoslibrary
    9156
```

It would be a bit of a slog for me :)

**Comment by RhetTbull** [comment]
Seems like a very useful feature. Currently, osxphotos doesn't even store keywords not associated with a photo but this can be accomplished with a query.  I'll plan to add a feature to `osxphotos keywords` to do this. In the meantime you can run this from the command line (assumes a Ventura database -- if you're on a different version, let me know as the query is slightly different depending on Photos version):

```bash
sqlite3 ~/Pictures/Photos\ Library.photoslibrary/database/Photos.sqlite "SELECT ZKEYWORD.ZTITLE FROM ZKEYWORD \
 LEFT JOIN Z_1KEYWORDS ON ZKEYWORD.Z_PK = Z_1KEYWORDS.Z_40KEYWORDS \
 LEFT JOIN ZADDITIONALASSETATTRIBUTES ON Z_1KEYWORDS.Z_1ASSETATTRIBUTES = ZADDITIONALASSETATTRIBUTES.Z_PK \
 LEFT JOIN ZASSET ON ZADDITIONALASSETATTRIBUTES.ZASSET = ZASSET.Z_PK \
 WHERE ZASSET.ZUUID IS NULL;"
```

**Comment by RhetTbull** [comment]
Copy the sqlite3  command and paste into Terminal to get a list of all keywords that aren't associated with a photo

**Comment by dmd** [comment]
Thank you, that worked. (I only had 8, all typos, much better than I feared.)

**Comment by dmd** [comment]
Actually, huh, this is interesting - *none* of those show up in the Manage My Keywords list. 

So it seems there's *three* levels of keywords:

- keywords used by a photo
- keywords used by no photos, which show in Manage My Keywords
- keywords used by no photos, which do not show in Manage My Keywords


What gives?

**Comment by RhetTbull** [comment]
BTW, I'm working on a new library for working with Photos ([photokit](https://github.com/RhetTbull/photokit)) -- I've figured out the private APIs Apple uses for Photos and I will be able to edit the database (something osxphotos doesn't do except in very specific circumstances).  The code is pre-alpha and incomplete but I can already open any library (not just system library), add/export assets, add/remove keywords, etc. When done this will enable several new osxphotos features (like copy between libraries) but it would also enable something like `osxphotos delete-unused-keywords` where the code could just delete the keywords from the database.

**Comment by dmd** [comment]
I trust you but I'd still be _extremely_ leery of having anything edit Apple's databases using a private API...

**Comment by RhetTbull** [comment]
> I trust you but I'd still be _extremely_ leery of having anything edit Apple's databases using a private API...

Yes, there's some risk there but these are the APIs that Photos itself uses...I'm not modifying the database directly, but calling the APIs to do it just as Photos would.  I've done extensive testing and so far things seem to be working well. Of course, because the APIs are private, it's entirely possible that Apple will break it with an OS upgrade (but that happens with osxphotos anyway...they change the Photos schema with each macOS version). 

As with any osxphotos code that changes the database, I'll have a scary warning about risk & ensuring a good backup as I do with `osxphotos timewarp`.  For example, `timewarp` currently prints this warning and prompts the user to confirm they want to proceed:

`**Caution**: This app directly modifies your Photos library database using undocumented features.  It may corrupt, damage, or destroy your Photos library.  Use at your own caution.  I strongly recommend you make a backup of your Photos library before using this script (e.g. use Time Machine).`

**Comment by RhetTbull** [comment]
> Actually, huh, this is interesting - none of those show up in the Manage My Keywords list.

Interesting. I'm not sure the cause. The ones that I got from the query did show up in the Keyword Manager.

Do you have Hidden photos? If so, these are normally excluded from searches. Is it possible the keywords are assigned to hidden photos?

I looked at the `ZKEYWORD` table and I see nothing that would indicate a separate status for "no matching photos".

**Comment by dmd** [comment]
Nope, no hidden photos.

**Comment by RhetTbull** [comment]
@dmd give v0.63.5 a try. `osxphotos keywords` under this version should list all keywords including those that have zero photos associated.

**Comment by RhetTbull** [comment]
@all-contributors please add @dmd for ideas

**Comment by dmd** [comment]
> @dmd give v0.63.5 a try. `osxphotos keywords` under this version should list all keywords including those that have zero photos associated.

Yep, that works!
It still however shows a few items that do NOT exist in the Manage My Keywords list.

```
  black an: 0
  dreaing: 0
  morris: 0
  palying: 0
  thiskeyworddoesnotexist: 0
```

For example, note that _thiskeyworddoesnotexist_ shows up:

<img width="486" alt="image" src="https://github.com/RhetTbull/osxphotos/assets/41439/dce438a5-1908-49c9-82dc-1ba1bb96a32b">


But _dreaing_ does not:

<img width="477" alt="image" src="https://github.com/RhetTbull/osxphotos/assets/41439/64019add-25c9-4137-b270-b79411ca3296">

**Comment by RhetTbull** [comment]
Try the following. Does the missing keyword show up in one of these?  Both commands will sort the keywords alphabetically to make it easier to check.

`sqlite3 ~/Pictures/Photos\ Library.photoslibrary/database/Photos.sqlite "SELECT ZTITLE FROM ZKEYWORD ORDER BY ZTITLE"`

Then run `osxphotos repl` and once the `>>>` prompt appears, type:

`sorted(photosdb._dbkeywords_keyword.keys())`

I don't understand what's going on as osxphotos now loads *every* keyword from the database and prints out the union of the `ZKEYWORD` table (Which contains every keyword in Photos) and the internal data structure (`_dbkeywords_keywords`) that osxphotos uses to track keyword to photo.

**Comment by RhetTbull** [comment]
Of course, I have no idea internally what Photos is doing but one idea may be that photos is showing only keywords that have at one time been assigned to a photo but not those that were created but never assigned. (They were typed but then deleted before being saved to the photo). One thing I've figured out from the private API is that the creation of a keyword object (which puts it into the ZKEYWORD table--I've verified this) is separate from assigning the keyword to a  photo. You must first create the keyword object (or fetch an existing one) then perform a change request on the photo to assign the keyword to the photo. Maybe there's a count somewhere in the database of how many times the keyword has been assigned.

**Comment by dmd** [comment]
Well this is interesting?

```
>>> 'palying' in photosdb._dbkeywords_keyword.keys()
False

>>>

$ sqlite3 ~/Pictures/Photos\ Library.photoslibrary/database/Photos.sqlite "SELECT ZTITLE FROM ZKEYWORD ORDER BY ZTITLE" | grep palying
palying
```

**Comment by RhetTbull** [comment]
> So, palying is in photosdb._dbkeywords_keyword but not in ZKEYWORD.ZTITLE

I think you mean the opposite, no? It's in `ZKEYWORD.ZTITLE` but not in `photosdb._dbkeywords_keyword `

That would indicate it is not assigned to any photo as `photosdb._dbkeywords_keyword` only has entries for keywords mapped to photo UUIDs.

Does it show up in the Keyword Manager?

**Comment by dmd** [comment]
Yes sorry that's what I meant. No, it does NOT show up in Keyword Manager.

**Comment by RhetTbull** [comment]
Interesting. Poking around the private APIs, there's a `requestZeroKeywordsWithOptions:(id)arg1 error:(id*)arg2;` method that might offer a way to get the "official" list of zero photo keywords rather than read directly from the database as osxphotos is currently doing. I'll play with this when I get  a chance to see if it does what I think it does.

**Comment by dmd** [comment]
My suspicion would be that that method supports "Keyword is not Any" in Smart Albums (which wasn't possible until the latest major rev of Photos). In other words, it's for getting photos with no keywords, not keywords with no photos.

**Comment by RhetTbull** [comment]
Well bummer. I tried it and it does nothing -- returns nil. It's quite possible I'm calling it wrong (these undocumented APIs take a lot of trial & error to figure out) but it didn't crash (the usual result of calling a private API incorrectly). I'm not sure what to tell you about the missing keyword but if it doesn't appear in Keyword Manager I guess it's not an issue.

**Comment by dmd** [comment]
Yeah, and more importantly they do not appear as autocomplete suggestions.
